### PR TITLE
Daniela - Leaves

### PR DESCRIPTION
**Pull request by dnsanche**
# Media Ranker
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
---      | ---
Describe a custom model method you wrote. | A custom method I wrote is the top 10. It accepts a parameter which is the category and then, it goes through all the database looking for elements that match this parameter. Then it sorts these elements by the number of votes in descending order. Finally, it returns the first 10 elements.
Describe how you approached testing that model method. What edge cases did you come up with? | I considered 3 cases, if there were no works of the category of the argument, if there were more than 10 elements, and when there were less than 10 elements.
What are `session` and `flash`? What is the difference between them? | Session allows you to track the user by the user_id and flash allows you to show messages when an action is performed.
What was one thing that you gained more clarity on through this assignment? | Model testing, flash, session, fixtures.
What is the Heroku URL of your deployed application? |  https://dani-media-ranker.herokuapp.com
Do you have any recommendations on how we could improve this project for the next cohort? |

### CODE REVIEW

File: app/views/works/index.html.erb [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+          <tbody>
+            <%Work.sort_desc("album").each do |work| %>
+              <tr>
+                <td width="80"><%=work.total_votes%></td>
+                <td><%=link_to work.title, work_path(work.id)%></td>
+                <td><%=work.creator%></td>  
+                <td><%=work.publication_year%></td> 
+                <td><%=link_to "Vote", work_votes_path(work.id), method: :post, :class=>"btn btn-primary"%></td>
+              </tr>          
+            <%end %>
```
Comment (dHelmgren) [inline app/views/works/index.html.erb:33]: This code is repeated 3 times. How could you use a view partial to DRY this up?

File: config/routes.rb [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+  resources :users, only: [:index, :new, :create, :show]
+
+  # get "/login", to: "users#login_form", as: "login"
+  # post "/login", to: "users#login"
+  # post "/logout", to: "users#logout", as: "logout"
+  get "/user/current", to: "users#current", as: "current_user"
+  
+  resources :works, only: [:show] do
+    resources :votes, only: [:create]
+  end
```
Comment (dHelmgren) [inline config/routes.rb:14]: You can do this with a single call to `resources :works`, because the `resources :votes, only: [:create]` covers the nested route

File: app/views/homepages/index.html.erb [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+          <h4>
+            <%= link_to book.title, work_path(book.id)%>
+            <small class="top-ten__creator"> by <%=book.creator%></small>         
+          </h4>
+          <p><%=book.total_votes%> votes</p>
+           </li>
+          <%end %>
+        </ul>
+        <!-- <a href="/works.books">View More Books</a> -->
+      </section>
```
Comment (dHelmgren) [inline app/views/homepages/index.html.erb:53]: This code gets repeated, making it a good spot for a view partial

File: app/controllers/users_controller.rb [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+  #     flash[:failure] = "Name cannot be blank"
+  #     redirect_to login_path
+  #   else
+  #     user = User.find_by(name: name)
+  #     if user
+  #       session[:user_id] = user.id
+  #       flash[:success] = "Successfully logged in as returning user #{name}"
+  #     else
+  #       user = User.create(name: name)
+  #       session[:user_id] = user.id
```
Comment (dHelmgren) [inline app/controllers/users_controller.rb:32]: Delete extended comments please!

File: test/models/work_test.rb [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+      creator: "Josh Howe", publication_year: 2019, description: "gsgdfg", total_votes: 1)
+  }
+  let (:new_work3) {
+    Work.new(category: "book", title: "3", 
+      creator: "Josh Howe", publication_year: 2019, description: "gsgdfg", total_votes: 1)
+  }
+  let (:new_work4) {
+    Work.new(category: "book", title: "4", 
+      creator: "Josh Howe", publication_year: 2019, description: "gsgdfg", total_votes: 1)
+  }
```
Comment (dHelmgren) [inline test/models/work_test.rb:19]: Are these getting used? This is what fixtures are for!

**Comment by dHelmgren** [comment]
# Media Ranker
## What We're Looking For



### Manual Testing

Workflow | yes / no
--- | ---
Deployed to Heroku | yes
Optional: Look and feel is similar to the original (consider styling and layout) | yes
**Before logging in** |
On index page, there are at most 10 pieces of media on three lists, and a Media Spotlight | yes
Can go into a work's show page | yes
Verify unable to vote on a work, and get a flash message | yes
Can edit this work successfully, and get a flash message | yes
Can go to "View all media" page and see three lists of works, sorted by vote | yes
Verify unable to create a new work when the form is empty, and details about the validation errors are visible to the user through a flash message | yes
Can create a new work successfully. Note the URL for this work's show page | yes
Can delete this work successfully | yes
Going back to the URL of this deleted work's show page produces a 404 or some redirect behavior (and does not try to produce a broken view) | yes
Verify that the "View all users" page lists no users (except what is produced by db seeding) | yes
**Log in** |
Logging in with a valid name changes the UI to "Logged in as" and "Logout" buttons | yes
Your username is listed in "View all users" page | yes
Verify that number of votes determines the Media Spotlight | yes
Voting on several different pieces of media affects the "Votes" tables shown in the work's show page and the user's show page | yes
Voting on the same work twice produces an error/flash message, and there is no extra vote | yes
**Log out** |
Logging out showed a flash message and changed the UI | yes
Logging in as a new user creates a new user | yes
Logging in as an already existing user has a specific flash message | yes

### Targeted Code Review

Area | yes / no
--- | ---
git commits were small and atomic, with useful messages | yes
Routes file uses `resources` for works | yes
Routes file shows intention in limiting routes for voting, log-in functionality, and users | yes
The homepage view, all media view, and new works view use semantic HTML | yes
The homepage view, all media view, and new works view use partials when appropriate | no, see comments
The model for media (likely named `work.rb`) `has_many` votes | yes
The model for media has methods to describe business logic, specifically for top ten and top media, possibly also for getting works by some category | yes
Some controller, likely the `ApplicationController`, has a controller filter for finding a logged in user | yes
Some controller, likely the `WorksController`, has a controller filter for finding a work | yes
The `WorksController` uses strong params | yes
The `WorksController`'s code style is clean, and focused on working with requests, responses, `params`, `session`, `flash` | yes

#### Targed Test Review

Area | yes / no
--- | ---
There are valid fixtures files used for users, votes, and works | yes
User model has tests with sections on validations (valid and invalid) and relationships (has votes) | no, commented out
Vote model has tests with sections on validations (valid and invalid) and relationships (belongs to a user, belongs to a vote) | no, commented out
Work model has tests with sections on validations (valid and invalid) and relationships (has votes) | yes, but incomplete
Work model has tests with a section on all business logic methods in the model, including their edge cases | yes, but incomplete

## Overall Feedback
This is generally a really solid project, but what's keeping it from being green is the commented out tests. Realistically, testing should be the first thing that you focus on with these apps, because it can catch so much of what might go wrong!
